vms/platformvm/block/builder/builder.go

```diff
@@ -241,8 +233,24 @@ func buildBlock(
 		)
 	}
 
+	var (
+		blockTxs      []*txs.Tx
+		remainingSize = targetBlockSize
+	)
+
+	for {
+		tx := builder.Mempool.Peek(remainingSize)
+		if tx == nil {
+			break
+		}
+		builder.Mempool.Remove([]*txs.Tx{tx})
+
+		remainingSize -= len(tx.Bytes())
+		blockTxs = append(blockTxs, tx)
```

Comment:
Next step is to verify each tx against a stateDiff before appending here- will be done in a follow-up.

Reply:
This seems like a pretty code piece to defer to a follow-up? If merged, it seems like that tag of dev would be unsafe/broken.

Reply:
q: it seems that Peek is only used here and it is followed by Remove. Should we instead have a Pop method that remove directly the tx and reinsert the tx if size limit is hit?
It may be marginally more memory efficient

Reply:
> This seems like a pretty code piece to defer to a follow-up? If merged, it seems like that tag of dev would be unsafe/broken.

What's done in this PR is a no-op change, the iteration is being moved from the mempool to the consumer of the mempool. We do not verify any of the txs before putting them into a block pre or post this PR

Reply:
> q: it seems that Peek is only used here and it is followed by Remove. Should we instead have a Pop method that remove directly the tx and reinsert the tx if size limit is hit?
It may be marginally more memory efficient

Can you come up with an implementation and benchmark it? I think the additional allocs for removing then re-adding would make it significantly slower

Reply:
With Peek+Remove we perform two operations,  lookup and remove, every time.
With Pop we would spare the lookup operation and just remove the tx from the mempool every time. Then we'd have a possible reinsert in case the mempool is so full that we have not included all txs in the block.
Isn't it guaranteed that the Pop is more effcient, albeit of a slight margin?

Reply:
The possible reinsert is my concern- I think it would be significantly slower. Post-BaseTx support, we could run this reinsert many times. Can you prepare a benchmark so we can evaluate empirically?

Reply:
The P-Chain mempool is frequently full of txs that cannot be fit into a single block. It's more likely that it is *not* full of txs:

<img width="1182" alt="image" src="https://github.com/ava-labs/avalanchego/assets/7675102/14461cae-932c-45e5-993a-6593f5fb5bfa">

Reply:
Re-add would break the FIFO ordering, so it's not equivalent to Peek-Remove